### DISCUSSION TITLE
Continuing the development of orcamap-react

### DISCUSSION THREAD

**Discussion by ayushman-t** [orcasound/orcamap-react#58] [Ideas]
# Project Inquiry

Hello, team!

I'm planning to continue the development of `orcamap-react`. I would appreciate your guidance on the project objectives and the best approach to planning its development.

Could you please advise on the following:

1. **Project Objectives**: What are the key objectives we need to achieve for the application?
   
2. **Development Plan**: How should we plan for its further development?

3. **Feature Prioritization**: Which features should be prioritized and implemented first?

Your insights and recommendations will be greatly appreciated. 

Thank you in advance!

Best regards,
Ayushman Tripathi

**Comment by scottveirs** [comment]
Hey @Ayushmanwebdeveloper thank you for posing the questions and re-starting the discussions. I'll respond briefly to each here and then as we make decisions, I'd like to work with you to distill any decisions into the [Wiki](https://github.com/orcasound/orcamap-react/wiki), README, and/or Issues organized/prioritized within a new [Project](https://github.com/orgs/orcasound/projects/40/views/2).

**1. Project objectives**

The vision of the orcamap project (running for a long time, like 2 decades!) is to build better-and-better visualizations and analyses of the movements, behavior, and phenology of the endangered Southern Resident killer whales. An emerging stretch goal is to also model and visualize predictions of where J, K, and L pods will move next (given a recent occurrence or real time track).

Here is a schematic that shows how orcamap could relate to other emerging apps (web or mobile) and the Acartia data cooperative:

![Simple 2022 SSEMMI_Acartia phase 2 schematic](https://github.com/orcasound/orcamap-react/assets/14044595/b73f6f2e-32d8-435f-8576-9a4cfb1a18ae)

Note that `orcamap.net` is envisioned as an output tool (highest piority), but it might make sense to also customize it as an input tool eventually under the assumption that extant reporting tools are usually based on species, or very rarely killer whale ecotypes (Southern Residents vs Bigg's) -- and it would be costly to add more granular identities to each of them (i.e. SRKW matrilines and individuals that are now commonly associated with community scientist sightings that include the ID-quality photos).

**2. Development plan**

Practically, we would like to deploy a web app via the reserved domain `orcamap.net` and a prototype of the web app has been deployed via Netlify. So, basic next steps include:

1. Get the current main branch working again
2. Merge or triage (archive) the [existing PR backlog](https://github.com/orcasound/orcamap-react/pulls)
4. Test that the CI with Netlify is still working
5. Obtain or rewrite the code from Sept 2022 hackathon participants that successfully pulled SRKW data from Acartia's API
6. Begin agile design/development of highest priority features

**3. Feature Prioritization**

The [fall 2022 hackathon Gdoc](https://docs.google.com/document/d/1Undj6mwUR2eHh-qO2Vk27O7mui4_VTIZbHMkolJ2H3Y/edit) has a good overview of the challenges we'd eventually like to surmount through the orcamap project or related efforts. I'd distill and reprioritize it now by focusing in from all species of interest to Orca Network to only occurrence data (acoustic or visual) for SRKWs. Whether to include the Bigg's ecotype as well -- now or later -- is a good question, but I think it can be tabled until we have a prototype working with SRKW data.

Highest priority: Refine existing basic Python or Javascript scripts to make Acartia.io API calls (e.g. pull SRKW rows for a month of interest) and visualization **point data** (e.g. a map of individual sightings displayed as icons). This was basically completed by [joyliao07](https://github.com/joyliao07) and [envm92](https://github.com/envm92) during the fall, 2022 hackathon, and separately by @cpsarason (see the [final slide deck from the 2022 Hackathon](https://docs.google.com/presentation/d/10s_mY7LmPnn3lckYNFJC2zJdD_1_xH8fF24eF0CK2kk/edit#slide=id.g12512f0f70d_0_0)) but could likely be refined to include parsing the `Comment` field to obtain pod occurrence (accomplished in a preliminary fashion by @liu-zoe in the [orca-salmon repo](https://github.com/salish-sea/orca-salmon) within the salish-sea organization. 

Secondary priority: Formulate a scheme for visualizing recent tracks (of at least SRKWs) from Acartia.io using open source visualization tools. @veirs worked on this in the 2022 hackathon and may have initial ideas for spatial and temporal grouping of points into tracks...
 
 A longer-term task I would like to tackle with help from you and other (present and past) contributors is to refine the roadmap -- probably in the new Project -- ensuring we don't miss any good ideas from previous iterations like the [original Trello roadmap for Orcamap](https://trello.com/b/MfDaW3P6/orcamap-development) and the [2021-2 Roadmap](https://github.com/orcasound/orcamap-react/issues/23) that @ivanoats drafted when he was mentoring a suite of prospective GSoC students.

**Reply by ayushman-t** [reply]
Hey! @scottveirs Many thanks for this. It was extremely helpful. I'll get to work on it, and hopefully we can work our way through the development plan's steps.

**Comment by scottveirs** [comment]
A high-level question raised by @paulcretu today in the Orcasound stand-up is whether to align on our selection of open mapping tools across projects. A cost of using different tools is difficulty of taking a cool new feature developed in one tool and replicating it in another tool. A benefit of using a variety of open mapping tools is that we can offer more new volunteers an opportunity to contribute -- no matter which mapping tool is their favorite.   

At the moment (Nov. 2023), we have a wide range of tools in play:

1. [Leaflet](https://leafletjs.com/) is used by v3 of the live-listening web app -- [live.orcasound.net](https://live.orcasound.net)
2. [Openlayers](https://openlayers.org/) is used by orcamap-react (mainly because we thought implementing custom features like movement models would be important) -- [orcamap.netlify.app/](https://orcamap.netlify.app/)
3. Mapbox is used by [Acartia.io](https://acartia.io). Before [Mapbox became not-fully open source](https://wptavern.com/mapbox-gl-js-is-no-longer-open-source), Christian (@cpsarason) also chose it for an [earlier version of orcamap](https://github.com/orcasound/orcamap) that allowed data input via a Mapbox+GoogleSheet integration.

Opinions @cpsarason or Ali @aalaydrus or Ivan @ivanoats ?

@paulcretu -- Can you remind me why Leaflet seemed the best choice for v3 of orcasite?

Looking further ahead, I'm anticipating at least one more map UI for the Salish Sea ecosystem (e.g. in this [2024 Acartia data flow schematic](https://docs.google.com/drawings/d/1kVTMfAF_hESYEjmJcOZ_ODC-6LnTb68_lKCcaw4jS6E/edit))... 

![2024 SSEMMI_Acartia phase 2 schematic](https://github.com/orcasound/orcamap-react/assets/14044595/55a83ac9-cdc0-418e-b4f6-e9824537df15)

and there probably will be many more -- like the one in the [orca-salmon dashboard](https://salmonorca.fly.dev/) that @liu-zoe implemented which also uses Mapbox.

**Reply by paulcretu** [reply]
> @paulcretu -- Can you remind me why Leaflet seemed the best choice for v3 of orcasite?

Just because it was the easiest to get working with React—[react-leaflet](https://react-leaflet.js.org/) is pretty good and really simple to integrate. Leaflet is missing a lot of advanced features, but I didn't need any of those features yet. OpenLayers seems great but takes a bit of wrangling to make it play nice with React's declarative rendering. Long term I was planning on switching to OpenLayers, once we need more advanced features (…when we add more layers?)

**Reply by paulcretu** [reply]
Also when I last looked, there didn't seem to be great options for React + OpenLayers libraries/wrappers. The solution reccomeneded on the OpenLayers website, [react-openlayers](https://github.com/allenhwkim/react-openlayers), hasn't been maintained in a while. But looks like there's a brand new library that may be worth trying out: https://planetlabs.github.io/maps/

**Reply by paulcretu** [reply]
One option could be to prototype advanced mapping features using OpenLayers in this repo, with the goal of eventually integrating into [orcasound/orcasite](https://github.com/orcasound/orcasite)

**Comment by cpsarason** [comment]
Hey all, apologies for the long silence here.

Scott, in answer to your question, I originally chose mapbox b.c. that is
what I was using at the time for another project and it integrated well.
They basically use leaflet as the core and then have a bunch of nice add
ons you can use.

In my more recent projects I have settled on leaflet, again mostly for
expediency....I integrated the Windy.com API with some of my own data and
they use leaflet. For those that don't know about windy, you should check
it out....it's a pretty great UI and really the best weather prediction app
that I've found.

My current stack is using Python on the backend to create geoJSON objects
that then gert rendered by the leaflet web page ... Again, mostly for
expediency, since I'm pulling in lots of disparate data feeds and using the
variety of Python tools to do so.

Make sense? Not sure what to recommend, but I have found leaflet to be
pretty robust so far.

Best
Christian

---
+1.206.992.1866
@cpsarason | cpsarason.github.io
